autogpt_platform/backend/backend/server/routers/integrations.py

```diff
@@ -80,20 +81,45 @@ async def callback(
     user_id: Annotated[str, Depends(get_user_id)],
     request: Request,
 ) -> CredentialsMetaResponse:
+    logger.debug(f"Received OAuth callback for provider: {provider}")
     handler = _get_provider_oauth_handler(request, provider)
 
     # Verify the state token
     if not await store.verify_state_token(user_id, state_token, provider):
+        logger.warning(f"Invalid or expired state token for user {user_id}")
         raise HTTPException(status_code=400, detail="Invalid or expired state token")
 
     try:
-        credentials = handler.exchange_code_for_tokens(code)
+        scopes = await store.get_scopes_from_state_token(user_id, state_token, provider)
+        logger.debug(f"Retrieved scopes from state token: {scopes}")
+
+        # If scopes are empty, use the default scopes for the provider
+        if not scopes:
+            logger.debug(f"Using default scopes for provider {provider}")
+            scopes = handler.DEFAULT_SCOPES
+
+        credentials = handler.exchange_code_for_tokens(code, scopes)
+        logger.debug(f"Received credentials with final scopes: {credentials.scopes}")
+
+        # Check if the granted scopes are sufficient for the requested scopes
+        if not set(scopes).issubset(set(credentials.scopes)):
+            # For now, we'll just log the warning and continue
+            logger.warning(
+                f"Granted scopes {credentials.scopes} for {provider}do not include all requested scopes {scopes}"
+            )
+
     except Exception as e:
-        logger.warning(f"Code->Token exchange failed for provider {provider}: {e}")
-        raise HTTPException(status_code=400, detail=str(e))
+        logger.error(f"Code->Token exchange failed for provider {provider}: {str(e)}")
```

Comment:
`f"{str(e)}"` is equivalent to `f"{e}"`
```suggestion
        logger.error(f"Code->Token exchange failed for provider {provider}: {e}")
```